flexget/plugins/api_tvmaze.py

```diff
@@ -0,0 +1,303 @@
+            session.add(db_episode)
+        db_episodes.append(db_episode)
+    return db_episodes
+
+
+def get_db_season(maze_id, seasons, session):
+    db_seasons = []
+    for season_num, season in seasons.items():
+        db_season = session.query(TVMazeSeasons).filter(and_(
+            TVMazeSeasons.series_maze_id == maze_id,
+            TVMazeSeasons.number == season_num)).first()
+        if not db_season:
+            db_season = TVMazeSeasons(maze_id=maze_id, season=season, session=session)
+            session.add(db_season)
+        db_seasons.append(db_season)
+    return db_seasons
+
+
+def get_db_genres(genres, session):
+    """Takes a list of genres as strings, returns the database instances for them."""
+    db_genres = []
+    for genre in genres:
+        db_genre = session.query(TVMazeGenre).filter(TVMazeGenre.name == genre).first()
+        if not db_genre:
+            db_genre = TVMazeGenre(name=genre)
+            session.add(db_genre)
+        db_genres.append(db_genre)
+    return db_genres
+
+
+def search_params_for_series(**lookup_params):
+    search_params = {
+        'maze_id': lookup_params.get('maze_id'),
+        'tvdb_id': lookup_params.get('tvdb_id'),
+        'tvrage_id': lookup_params.get('tvrage_id'),
+        'name': lookup_params.get('title')
+    }
+    return search_params
+
+
+@with_session
+def from_cache(session=None, search_params=None, cache_type=None):
+    """
+    Returns a result from requested table based on search params
+    :param session: Current session
+    :param search_params: Relevant search params. Should match table column names
+    :param cache_type: Object for search
+    :return: Query result
+    """
+    result = None
+    if not any(search_params.values()):
+        raise LookupError('No parameters sent for lookup')
+    else:
+        result = session.query(cache_type).filter(
+            or_(getattr(cache_type, col) == val for col, val in search_params.iteritems() if val)).first()
+    return result
+
+
+@with_session
+def from_lookup(session=None, title=None):
+    return session.query(TVMazeLookup).filter(func.lower(TVMazeLookup.search_name) == title.lower()).first()
+
+
+def prepare_lookup(**lookup_params):
+    """
+    Return a dict of params which is valid with pytvmaze get_show method
+    """
+    prepared_params = {}
+    series_name = lookup_params.get('series_name', lookup_params.get('show_name'))
+    year_match = re.search('\(([\d]{4})\)', series_name)  # Gets year from title if present
+    if year_match:
+        year_match = year_match.group(1)
+
+    prepared_params['maze_id'] = lookup_params.get('tvmaze_id')
+    prepared_params['tvdb_id'] = lookup_params.get('tvdb_id') or lookup_params.get('trakt_series_tvdb_id')
+    prepared_params['tvrage_id'] = lookup_params.get('tvrage_id') or lookup_params.get('trakt_series_tvrage_id')
+    prepared_params['show_name'] = re.sub('\(([\d]{4})\)', '', series_name).rstrip()  # Remove year from name if present
```

Comment:
cool. maybe we should use some of these method as global tools or something, in flexget.utils.tools